src/utilities.jl

```diff
@@ -0,0 +1,89 @@
+
+- `chain`: a chain object from MCMCChains
+- `data`: a vector of data used to estimate parameters of the model 
+- `ll_fun`: a function that computes the log likelihood of a single data point: f(θ1, ..., θn, data), where θi is the ith parameter
+"""
+function pointwise_loglikes(chain::Chains, data, ll_fun)
+    samples = Array(Chains(chain, :parameters).value)
+    pointwise_loglikes(samples, data, ll_fun)
+end
+
+"""
+    pointwise_loglikes(samples::Array{Float64,3}, data, ll_fun)
+
+Computes the pointwise log likelihood 
+
+- `samples`: a three dimensional array of mcmc samples where the first dimension represents the samples, the second dimension represents
+the samples, and the third dimension represents the chains. 
+- `data`: a vector of data used to estimate parameters of the model 
+- `ll_fun`: a function that computes the log likelihood of a single data point: f(θ1, ..., θn, data), where θi is the ith parameter
+"""
+function pointwise_loglikes(samples::Array{Float64,3}, data, ll_fun)
+    fun = (p,d)-> ll_fun(p..., d)
+    n_data = length(data)
+    n_samples, n_chains = size(samples)[[1,3]]
+    pointwise_lls = fill(0.0, n_data, n_samples, n_chains)
+    for c in 1:n_chains 
+        for s in 1:n_samples
+            for d in 1:n_data
+                pointwise_lls[d,s,c] = fun(samples[s,:,c], data[d])
+            end
+        end
+    end
+    return pointwise_lls
+end
+
+"""
+    compute_loo(psis_output, pointwise_lls)
+
+Computes leave one out (loo) cross validation based on posterior distribution of model parameters.
+
+- `pointwise_lls`: pointwise log likelihood where `pointwise_lls[d,s,c]` corresponds to log likelihood of 
+evaluated at datapoint d, sample s, for chain c.  
+"""
+function compute_loo(pointwise_lls)
+    psis_output = psis(pointwise_lls)
+    return compute_loo(psis_output, pointwise_lls)
+end
+
+"""
+    compute_loo(psis_output, pointwise_lls)
+
+Computes leave one out (loo) cross validation based on posterior distribution of model parameters.
+
+- `psis_output`: object returned from `psis`
+- `pointwise_lls`: pointwise log likelihood where `pointwise_lls[d,s,c]` corresponds to log likelihood of 
+evaluated at datapoint d, sample s, for chain c.  
+"""
+function compute_loo(psis_output, pointwise_lls)
+    dims = size(pointwise_lls)
+    lwp = deepcopy(pointwise_lls)
+    lwp += psis_output.weights
+    lwpt = reshape(lwp, dims[1], dims[2] * dims[3])'
+    loos = reshape(logsumexp(lwpt; dims=1), size(lwpt, 2))
+    return sum(loos)
+end
+
+
+"""
+    compute_loo(chain::Chain, data, ll_fun)
+
+Computes leave one out (loo) cross validation based on posterior distribution of model parameters.
+
+- `chain`: a chain object from MCMCChains
+- `data`: a vector of data used to estimate parameters of the model 
+- `ll_fun`: a function that computes the log likelihood of a single data point: f(θ1, ..., θn, data), where θi is the ith parameter
+"""
+function compute_loo(chain::Chains, data, ll_fun::Function)
```

Comment:
See above -- the function should call the method that's already been implemented, which provides more information (Like Monte Carlo standard errors, effective sample sizes, and Pareto k values).

Reply:
Ok. I did not realize that `psis_loo` was created. If I understand correctly, all that is needed is a way to extract or compute the pointwise log likelihoods, which can be passed to `psis_loo`. Is that true?

Reply:
And the Psisloo object returned contains the Psis object!

Reply:
Part of the problem is that I am not very familiar with psis. The other part is that `psisi_loo` has not documentation and outputs a lot of information into the REPL. So it is not currently clear to the user what the function does and how to navigate the output.

Reply:
> Part of the problem is that I am not very familiar with psis. The other part is that `psisi_loo` does not have documentation and outputs a lot of information into the REPL. So it is not currently clear to the user what the function does and how to navigate the output.

Yeah, completely understand/agree, I'll document it today.